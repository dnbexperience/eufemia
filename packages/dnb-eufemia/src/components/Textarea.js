/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index textarea to autogenerate all the components and patterns
 * Used by "prepareTextareas"
 */

import Textarea from './textarea/Textarea'
export * from './textarea/Textarea'
export default Textarea
