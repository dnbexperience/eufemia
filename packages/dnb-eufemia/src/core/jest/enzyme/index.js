/**
 * Default Component Test Setup
 *
 */

// enzyme setup
const enzyme = require('enzyme')

// TODO: This may not be needed in future
const Adapter = require('enzyme-adapter-react-16')
enzyme.configure({ adapter: new Adapter() })

module.exports = enzyme
