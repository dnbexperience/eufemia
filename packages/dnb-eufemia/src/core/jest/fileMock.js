export default 'static file content'
