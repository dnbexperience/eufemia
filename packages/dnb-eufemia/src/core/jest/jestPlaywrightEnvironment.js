/**
 * Jest Setup for Screenshot testing
 *
 */

const PlaywrightEnvironment =
  require('jest-playwright-preset/lib/PlaywrightEnvironment').default
const chalk = require('chalk')
const { config } = require('./jestSetupScreenshots')

class JestEnvironment extends PlaywrightEnvironment {
  constructor(config, context) {
    super(config, context)
  }

  async setup() {
    await super.setup()

    // jest-playwright exposes these on the global
    const browser = this.global.browser
    const browserName = this.global.browserName || 'firefox'

    if (browser && typeof browser.version === 'function') {
      const version = browser.version() // e.g. "129.0"
      // One neat line before your test output:
      console.log(`\nâ–¶ Using ${browserName} ${version}\n`)
    }
  }

  async teardown() {
    await super.teardown()
  }

  setRetryAttempt(retryAttempt) {
    this.global.retryAttempt = retryAttempt
  }

  getCurrentTestName(state) {
    const { currentlyRunningTest } = state
    return getParents(currentlyRunningTest).reverse().join(' ')
  }

  reportedPageUrl = []

  async handleTestEvent(event, state) {
    if (
      this.global.pageUrl &&
      !this.reportedPageUrl.includes(this.global.pageUrl)
    ) {
      this.reportedPageUrl.push(this.global.pageUrl)

      const cliColors = {
        reset: '\x1b[0m',
        bold: '\x1b[1m',
        dim: '\x1b[2m',
        yellow: '\x1b[33m',
        red: '\x1b[31m',
        green: '\x1b[32m',
        hidden: '\x1b[8m',
      }

      const themeName = this.global.themeName
        ? `${cliColors.bold}${this.global.themeName}${cliColors.reset} `
        : ''
      console.log(
        `${cliColors.yellow}URL:`,
        `${cliColors.reset}${themeName}${this.global.pageUrl}`
      )
    }

    if (config.retryTimes > 0) {
      // Reset retry attempt at the start of each test
      if (event.name === 'test_start') {
        this.setRetryAttempt(0)
      }

      if (event.name === 'test_fn_failure') {
        const currentTestName = this.getCurrentTestName(state)
        const retryAttempt = state.currentlyRunningTest.invocations
        this.setRetryAttempt(retryAttempt)

        // Set a marker on the page to indicate this is a retry
        if (this.global.page) {
          this.global.page
            .evaluate(() => {
              window.__VISUAL_TEST_RETRY__ = true
            })
            .catch(() => {
              // Ignore errors if page is not available
            })
        }

        console.log(
          chalk.yellow(
            `Retry attempt #${retryAttempt}: ${currentTestName}`
          )
        )
      }

      if (event.name === 'test_fn_success') {
        // Reset retry attempt on success
        this.setRetryAttempt(0)
      }
    }
  }
}

module.exports = JestEnvironment

function getParents(item) {
  const names = []

  if (item?.name) {
    names.push(item.name)
  }

  for (const key in item) {
    if (item[key]?.name) {
      names.push(item[key].name)
    }
    if (item[key]?.parent) {
      names.push(...getParents(item[key].parent))
    }
  }

  return names.filter(
    (name) => name !== 'ROOT_DESCRIBE_BLOCK' && name !== 'Error'
  )
}
