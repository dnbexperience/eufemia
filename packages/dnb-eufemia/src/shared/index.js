/**
 * Shared
 *
 */

import Context from './Context'
import Provider from './Provider'
import MediaQuery from './MediaQuery'
import useMediaQuery from './useMediaQuery'
import useMedia from './useMedia'

export { Context, Provider, MediaQuery, useMediaQuery, useMedia }

// deprecated can be removed in v10
export * from '../components/space/SpacingHelper'
