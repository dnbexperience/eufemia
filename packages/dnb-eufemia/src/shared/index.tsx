/**
 * Shared
 *
 */

export { default as Context } from './Context'
export { default as Provider } from './Provider'
export { default as Theme } from './Theme'
export { default as useTheme } from './useTheme'
export { default as VisibilityByTheme } from './VisibilityByTheme'
export { default as MediaQuery } from './MediaQuery'
export { default as useMediaQuery } from './useMediaQuery'
export { default as useMedia } from './useMedia'
export { default as useTranslation } from './useTranslation'
export { default as Translation } from './Translation'
export * from './Translation'

export type { ThemeNames } from './Theme'
