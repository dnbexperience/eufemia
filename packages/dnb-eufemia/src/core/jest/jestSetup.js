/**
 * Jest Setup
 *
 */

import { axe, toHaveNoViolations } from 'jest-axe'
import fakeProps from 'react-fake-props'
import { shallow, mount, render } from './enzyme'
import ReactDOMServer from 'react-dom/server'
import fs from 'fs-extra'
import onceImporter from 'node-sass-once-importer'
import path from 'path'
import sass from 'node-sass'
import { toBeType } from 'jest-tobetype'
import toJson from 'enzyme-to-json'

export {
  fakeProps, // we have also our own replacement function called "fakeAllProps"
  shallow,
  mount,
  render,
  toJson,
  axe,
  toHaveNoViolations
}

global.IS_TEST = true
if (typeof window !== 'undefined') {
  window.IS_TEST = true
}

expect.extend({ toBeType })
expect.extend(toHaveNoViolations)

export const loadScss = (file, options = {}) => {
  try {
    const sassResult = sass.renderSync({
      file,
      includePaths: [path.resolve(__dirname, '../../style/core/')],
      importer: [onceImporter()],
      ...options
    })
    return String(sassResult.css)
  } catch (e) {
    console.error('loadScss error:', e)
    return e
  }
}

export const mockGetSelection = () => {
  let memory
  Object.defineProperty(window.navigator, 'clipboard', {
    configurable: true,
    value: {
      writeText: jest.fn().mockImplementation((v) => {
        memory = v
        return Promise.resolve(v)
      }),
      readText: jest.fn().mockImplementation(() => Promise.resolve(memory))
    }
  })

  const mockRange = new (class Range {
    constructor() {
      this.startContainer = {
        parentNode: document.createElement('div')
      }
    }
    getElement() {
      return this.node
    }
    insertNode(elem) {
      this.node = document.createElement('div')
      this.node.appendChild(elem)
      return this
    }
    cloneRange() {
      return this
    }
  })()

  let ranges = [mockRange]

  const mockValue = '1234.56'
  let value = mockValue
  let rangeCount = 9

  class RangeObj {
    rangeCount = rangeCount
    toString = () => value
    addRange(range = mockRange) {
      value = mockValue
      ranges.push(range)
      rangeCount = ranges.length
    }
    getRangeAt(index) {
      return ranges[index]
    }
    removeAllRanges() {
      value = ''
      ranges = []
      rangeCount = ranges.length
    }
  }

  Object.defineProperty(window, 'getSelection', {
    configurable: true,
    value: () => {
      return new RangeObj()
    }
  })
}

export const loadImage = async (imagePath) =>
  await fs.readFile(path.resolve(imagePath))

export const toHtml = (Component) =>
  ReactDOMServer.renderToStaticMarkup(Component)

export const axeComponent = async (...components) => {
  const html = components
    .filter((Component) =>
      // enzyme names the mounted wrapper: ReactWrapper
      /react/i.test(String(Component.constructor))
    )
    .map((Component) => toHtml(Component))
    .join('\n')

  return await axe(
    `<main>${html}</main>`,
    typeof components[1] === 'object' ? components[1] : null
  )
}

export function attachToBody() {
  let container = document.createElement('div')
  document.body.append(container)
  return container
}

function jestSetup() {}

export default jestSetup
