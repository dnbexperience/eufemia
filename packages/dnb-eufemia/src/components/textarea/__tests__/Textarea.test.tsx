/**
 * Textarea Test
 *
 */

import { fireEvent, render, waitFor } from '@testing-library/react'
import React from 'react'
import { axeComponent, loadScss } from '../../../core/jest/jestSetup'
import Textarea, { TextareaProps } from '../Textarea'
import userEvent from '@testing-library/user-event'
import { Provider } from '../../../shared'

const props: TextareaProps = {
  id: 'textarea',
  label: null,
  status: null, // to make sure we don't get aria-details
  textarea_element: null,
  disabled: false,
}

describe('Textarea component', () => {
  it('has correct state after "focus" trigger', () => {
    render(
      <Textarea {...props} value={null}>
        {null}
      </Textarea>
    )
    fireEvent.focus(document.querySelector('textarea'))

    expect(document.querySelector('.dnb-textarea').classList).toContain(
      'dnb-textarea--focus'
    )
  })

  it('has correct state after "change" trigger', () => {
    render(
      <Textarea {...props} value={null}>
        {null}
      </Textarea>
    )
    expect(
      document.querySelector('.dnb-textarea').classList
    ).not.toContain('dnb-textarea--has-content')

    const value = 'new value'

    fireEvent.change(document.querySelector('textarea'), {
      target: { value },
    })

    expect(document.querySelector('.dnb-textarea').classList).toContain(
      'dnb-textarea--has-content'
    )
    expect(document.querySelector('textarea').value).toBe(value)
  })

  // // make sure getDerivedStateFromProps works
  it('has correct state after changing "value" prop (set by getDerivedStateFromProps)', () => {
    const { rerender } = render(
      <Textarea {...props} value={null}>
        {null}
      </Textarea>
    )
    const initValue = 'new prop value'
    const emptyValue = null

    rerender(
      <Textarea {...props} value={initValue}>
        {null}
      </Textarea>
    )
    expect(document.querySelector('textarea').value).toBe(initValue)

    rerender(
      <Textarea {...props} value={emptyValue}>
        {null}
      </Textarea>
    )
    expect(document.querySelector('textarea').value).toBe('')
  })

  it('events gets emitted correctly: "on_change" and "on_key_down"', async () => {
    const initValue = 'init value'
    const newValue = 'new value'

    const on_change = jest.fn()
    const on_key_down = jest.fn() // additional native event test

    render(
      <Textarea
        {...props}
        value={initValue}
        on_change={on_change}
        on_key_down={on_key_down} // additional native event test
      />
    )

    expect(document.querySelector('textarea').value).toBe(initValue)
    userEvent.type(document.querySelector('textarea'), newValue)
    await waitFor(() => {
      expect(on_change.mock.calls.length).toBe(9)
      expect(document.querySelector('textarea').value).toBe(
        initValue + newValue
      )
    })

    // additional native event test
    fireEvent.keyDown(document.querySelector('textarea'), {
      keyCode: 84, // space,
      key: 'Space',
    })
    await waitFor(() => {
      expect(on_key_down.mock.calls.length).toBe(10)
      expect(on_key_down.mock.calls[0][0].rows).toBe(1)
    })
  })

  it('supports null as value', () => {
    render(<Textarea {...props} value={null} />)

    expect(document.querySelector('textarea').value).toBe('')
  })

  it('has correct state after setting "value" prop using placeholder (set by getDerivedStateFromProps)', () => {
    const { rerender } = render(<Textarea placeholder="Placeholder" />)

    const newValue = 'new value'
    const emptyValue = null
    const zeroValue = '0'

    rerender(<Textarea value={newValue} />)
    expect(document.querySelector('textarea').value).toBe(newValue)

    rerender(<Textarea value={emptyValue} />)
    expect(document.querySelector('textarea').value).toBe('')

    rerender(<Textarea value={zeroValue} />)
    expect(document.querySelector('textarea').value).toBe(
      String(zeroValue)
    )
  })

  it('uses children as the value', () => {
    render(<Textarea>children</Textarea>)
    expect(document.querySelector('textarea').value).toBe('children')
  })

  it('has correct size attribute (chars length) on textarea by using textarea_attributes', () => {
    render(<Textarea textarea_attributes={{ size: 2 }} />)
    expect(document.querySelector('textarea').getAttribute('size')).toBe(
      '2'
    )
  })

  it('has to to have a prop value like value', () => {
    const { rerender } = render(
      <Textarea {...props} value={null}>
        {null}
      </Textarea>
    )
    const value = 'new value'
    rerender(
      <Textarea {...props} value={value}>
        {null}
      </Textarea>
    )
    expect(document.querySelector('textarea').value).toBe(value)
  })

  it('has to to have a label value as defined in the prop', () => {
    render(<Textarea {...props} label="label" />)
    expect(document.querySelector('label').textContent).toBe('label')
  })

  it('has to to have a status value as defined in the prop', () => {
    render(<Textarea {...props} status="status" status_state="error" />)
    expect(
      document.querySelector('.dnb-form-status__text').textContent
    ).toBe('status')
  })

  it('has a disabled attribute, once we set disabled to true', () => {
    const { rerender } = render(<Textarea />)
    rerender(<Textarea disabled={true} />)
    expect(document.querySelector('textarea')).toHaveAttribute('disabled')
  })

  it('will correctly auto resize if prop autoresize is used', async () => {
    render(<Textarea rows={1} autoresize={true} autoresize_max_rows={4} />)

    const elem = document.querySelector('textarea')

    const style = {
      lineHeight: String(1.5 * 16),
    } as CSSStyleDeclaration

    jest.spyOn(window, 'getComputedStyle').mockImplementation(() => style)

    jest
      .spyOn(elem, 'scrollHeight', 'get')
      .mockImplementation(() => 1.5 * 16)
    userEvent.type(elem, 'a')
    await waitFor(() => {
      expect(elem.style.height).toBe('24px')
    })

    jest
      .spyOn(elem, 'scrollHeight', 'get')
      .mockImplementation(() => 1.5 * 32)
    userEvent.type(elem, 'a')
    await waitFor(() => {
      expect(elem.style.height).toBe('48px')
    })

    jest
      .spyOn(elem, 'scrollHeight', 'get')
      .mockImplementation(() => 1.5 * 2000)
    userEvent.type(elem, 'a')
    await waitFor(() => {
      expect(elem.style.height).toBe('96px')
    })
  })

  it('should support spacing props', () => {
    render(<Textarea top="2rem" />)

    const element = document.querySelector('.dnb-textarea')

    expect(Array.from(element.classList)).toEqual([
      'dnb-textarea',
      'dnb-textarea--virgin',
      'dnb-form-component',
      'dnb-space__top--large',
    ])
  })

  it('should inherit formElement vertical label', () => {
    render(
      <Provider formElement={{ label_direction: 'vertical' }}>
        <Textarea label="Label" />
      </Provider>
    )

    const element = document.querySelector('.dnb-textarea')
    const attributes = Array.from(element.attributes).map(
      (attr) => attr.name
    )

    expect(attributes).toEqual(['class'])
    expect(Array.from(element.classList)).toEqual([
      'dnb-textarea',
      'dnb-textarea--virgin',
      'dnb-form-component',
      'dnb-textarea--vertical',
    ])
  })

  it('should validate with ARIA rules as a textarea with a label', async () => {
    const Comp = render(
      <>
        <label htmlFor="textarea">text</label>
        <Textarea {...props} id="textarea" value="some value" />
      </>
    )

    expect(await axeComponent(Comp)).toHaveNoViolations()
  })

  it('gets valid ref element', () => {
    let ref: React.RefObject<HTMLTextAreaElement>

    function MockComponent() {
      ref = React.useRef()
      return <Textarea {...props} inner_ref={ref} />
    }

    render(<MockComponent />)

    expect(ref.current.classList).toContain('dnb-textarea__textarea')
    expect(ref.current.tagName).toBe('TEXTAREA')
    expect(ref.current).toBeInstanceOf(HTMLTextAreaElement)
  })

  it('gets valid element when ref is function', () => {
    const ref: React.MutableRefObject<HTMLTextAreaElement> =
      React.createRef()

    const refFn = (elem: HTMLTextAreaElement) => {
      ref.current = elem
    }

    render(<Textarea id="unique" inner_ref={refFn} />)

    expect(ref.current.getAttribute('id')).toBe('unique')
    expect(ref.current.classList).toContain('dnb-textarea__textarea')
    expect(ref.current.tagName).toBe('TEXTAREA')
    expect(ref.current).toBeInstanceOf(HTMLTextAreaElement)
  })

  it('should render characterCounter', async () => {
    const { rerender } = render(
      <Textarea characterCounter={{ max: 8 }} value="foo" />
    )

    const counter = document.querySelector('.dnb-text-counter__message')
    const textarea = document.querySelector('textarea')
    const ariaLive = document.querySelector('.dnb-aria-live')

    expect(counter).toHaveTextContent('3 av 8 tegn gjenstår')
    expect(ariaLive).toHaveTextContent('')

    await userEvent.type(textarea, 'bar')

    expect(counter).toHaveTextContent('6 av 8 tegn gjenstår')
    expect(ariaLive).toHaveTextContent('6 av 8 tegn gjenstår')

    rerender(
      <Textarea characterCounter={{ max: 8 }} value="foo" lang="en-GB" />
    )

    expect(counter).toHaveTextContent('6 of 8 characters remaining')

    await userEvent.type(textarea, 'baz')

    expect(ariaLive).toHaveTextContent('8 of 8 characters remaining')

    rerender(
      <Textarea
        characterCounter={{ max: 8, variant: 'up' }}
        value="foo"
        lang="en-GB"
      />
    )

    expect(counter).toHaveTextContent('You have used 8 of 8 characters')
  })
})

describe('Textarea scss', () => {
  it('has to match style dependencies css', () => {
    const css = loadScss(require.resolve('../style/deps.scss'))
    expect(css).toMatchSnapshot()
  })

  it('have to match default theme snapshot', () => {
    const css = loadScss(
      require.resolve('../style/themes/dnb-textarea-theme-ui.scss')
    )
    expect(css).toMatchSnapshot()
  })
})
