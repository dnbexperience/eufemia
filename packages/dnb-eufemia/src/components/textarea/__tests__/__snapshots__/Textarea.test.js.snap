// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea component have to match snapshot 1`] = `
<Textarea
  0={
    {
      "displayName": "Textarea",
      "props": {
        "align": "'left'",
        "autoresize": "autoresize",
        "autoresize_max_rows": "autoresize_max_rows",
        "children": "children",
        "class": "class",
        "className": "className",
        "cols": 1,
        "custom_element": {},
        "custom_method": [Function],
        "disabled": "disabled",
        "global_status_id": "global_status_id",
        "id": "id",
        "inner_ref": [Function],
        "label": "label",
        "label_direction": "'horizontal'",
        "label_sr_only": "label_sr_only",
        "on_blur": [Function],
        "on_change": [Function],
        "on_focus": [Function],
        "on_key_down": [Function],
        "on_state_update": [Function],
        "placeholder": "placeholder",
        "readOnly": "readOnly",
        "rows": 1,
        "skeleton": "skeleton",
        "status": "status",
        "status_no_animation": "status_no_animation",
        "status_props": {},
        "status_state": "status_state",
        "stretch": "stretch",
        "suffix": "suffix",
        "textarea_attributes": "textarea_attributes",
        "textarea_class": "textarea_class",
        "textarea_element": [Function],
        "textarea_state": "textarea_state",
        "value": "value",
      },
    }
  }
  align={null}
  autoresize={null}
  autoresize_max_rows={null}
  class={null}
  className={null}
  cols={null}
  custom_element={null}
  custom_method={null}
  disabled={false}
  global_status_id={null}
  id="textarea"
  inner_ref={null}
  label={null}
  label_direction={null}
  label_sr_only={null}
  on_blur={null}
  on_change={null}
  on_focus={null}
  on_key_down={null}
  on_state_update={null}
  placeholder={null}
  readOnly={false}
  rows={null}
  skeleton={null}
  status={null}
  status_no_animation={null}
  status_props={null}
  status_state="error"
  stretch={null}
  suffix={null}
  textarea_attributes={null}
  textarea_class={null}
  textarea_element={null}
  textarea_state={null}
  value="test"
>
  <span
    className="dnb-textarea dnb-textarea--virgin dnb-form-component dnb-textarea--has-content"
  >
    <span
      className="dnb-textarea__inner"
    >
      <AlignmentHelper
        className={null}
      >
        <span
          aria-hidden={true}
          className="dnb-alignment-helper"
        />
      </AlignmentHelper>
      <FormStatus
        attributes={null}
        class={null}
        className={null}
        global_status_id={null}
        icon="error"
        icon_size="medium"
        id="textarea-form-status"
        label={null}
        no_animation={null}
        role={null}
        show={null}
        size="default"
        skeleton={null}
        state="error"
        stretch={null}
        text={null}
        text_id="textarea-status"
        title={null}
        variant={null}
        width_element={null}
        width_selector={null}
      />
      <span
        className="dnb-textarea__row"
      >
        <span
          className="dnb-textarea__shell"
        >
          <textarea
            className="dnb-textarea__textarea"
            disabled={false}
            id="textarea"
            name="textarea"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="textbox"
            value="test"
          />
          <span
            className="dnb-textarea__state"
          />
        </span>
      </span>
    </span>
  </span>
</Textarea>
`;

exports[`Textarea scss have to match default theme snapshot 1`] = `
"/*
* Textarea theme
*
*/
/*
 * Utilities
 */
:root {
  --textarea-rows: 2;
}

.dnb-textarea__textarea {
  color: var(--color-black);
  background-color: var(--color-white);
  overflow-y: auto;
  overscroll-behavior: contain;
  scrollbar-width: thin;
  -webkit-overflow-scrolling: touch;
  -ms-overflow-style: auto;
  scrollbar-color: var(--color-sea-green) transparent;
}
.dnb-textarea__textarea ::selection {
  background-color: var(--color-mint-green);
  color: var(--color-black);
  text-shadow: none;
}
@supports not (scrollbar-color: auto) {
  .dnb-textarea__textarea::-webkit-scrollbar {
    background-color: var(--color-black-8);
  }
  .dnb-textarea__textarea::-webkit-scrollbar:vertical {
    width: 0.5rem;
  }
  .dnb-textarea__textarea::-webkit-scrollbar:horizontal {
    height: 0.5rem;
  }
  .dnb-textarea__textarea::-webkit-scrollbar-thumb {
    background-color: var(--color-sea-green);
    border-radius: 0.5rem;
  }
  .dnb-textarea__textarea::-webkit-scrollbar-thumb:hover {
    background-color: var(--color-emerald-green);
  }
}
html:not([data-visual-test]) .dnb-textarea__textarea {
  scroll-behavior: smooth;
}
html[data-visual-test] .dnb-textarea__textarea {
  scroll-behavior: auto !important;
}
.dnb-textarea__placeholder {
  color: var(--color-black-55);
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: var(--textarea-rows, 2);
}
.dnb-textarea__textarea:not([disabled]):not(.dnb-textarea--disabled):focus ~ .dnb-textarea__state, .dnb-textarea__textarea:not([disabled]):not(.dnb-textarea--disabled):hover ~ .dnb-textarea__state {
  box-shadow: 0 0 0 0.125rem var(--color-emerald-green);
}
.dnb-textarea--disabled .dnb-textarea__textarea, .dnb-textarea__textarea[disabled] {
  user-select: none;
  -webkit-user-select: none;
  color: var(--color-black-55);
  background-color: var(--color-black-3);
}
.dnb-textarea--disabled .dnb-textarea__textarea::-webkit-scrollbar, .dnb-textarea__textarea[disabled]::-webkit-scrollbar {
  width: 0;
}
.dnb-textarea--disabled .dnb-textarea__state, .dnb-textarea__textarea[disabled] ~ .dnb-textarea__state {
  box-shadow: 0 0 0 0.0625rem var(--color-black-55);
  background-color: var(--color-black-3);
}
.dnb-textarea__status--error:not(.dnb-textarea--focus) .dnb-textarea__textarea:not([disabled]):not(.dnb-textarea--disabled), .dnb-textarea__status--error:not(.dnb-textarea--focus) .dnb-textarea__textarea:not([disabled]):not(.dnb-textarea--disabled):hover {
  color: var(--color-fire-red);
}
.dnb-textarea__status--error:not(.dnb-textarea--focus) .dnb-textarea__textarea:not([disabled]):not(.dnb-textarea--disabled) ~ .dnb-textarea__state {
  --border-color: var(--color-fire-red);
  box-shadow: 0 0 0 0.0625rem var(--border-color);
  border-color: transparent;
}
.dnb-textarea__status--error:not(.dnb-textarea--focus) .dnb-textarea__textarea:not([disabled]):not(.dnb-textarea--disabled):hover ~ .dnb-textarea__state {
  --border-color: var(--color-fire-red);
  box-shadow: 0 0 0 0.125rem var(--border-color);
  border-color: transparent;
}
.dnb-textarea--has-content .dnb-textarea__textarea ~ .dnb-textarea__placeholder, .dnb-textarea--focus .dnb-textarea__textarea:not([disabled]):not([readonly]) ~ .dnb-textarea__placeholder {
  display: none;
}"
`;

exports[`Textarea scss have to match snapshot 1`] = `
"/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
 * FormLabel component
 *
 */
/*
 * Utilities
 */
.dnb-form-label {
  display: inline-block;
  width: auto;
  padding: 0;
  margin-right: 1rem;
  text-align: left;
  vertical-align: baseline;
  white-space: pre-wrap;
  word-break: normal;
  font-size: var(--font-size-basis);
  color: inherit;
}
.dnb-form-label--vertical {
  display: block;
  margin-right: 0;
  margin-bottom: 0.5rem;
}
.dnb-form-label[for]:not([disabled]) {
  user-select: none;
  -webkit-user-select: none;
  cursor: pointer;
}
.dnb-form-label[disabled] {
  cursor: not-allowed;
}

legend.dnb-form-label {
  display: inline-block;
}

/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
 * FormStatus component
 *
 */
/*
 * Utilities
 */
:root {
  --form-status-radius: 0.25rem;
}

.dnb-form-status {
  display: flex;
  opacity: 1;
  transition: height 400ms var(--easing-default), opacity 400ms var(--easing-default), margin 400ms var(--easing-default), padding 400ms var(--easing-default);
}
.dnb-form-status--hidden {
  will-change: height, opacity, margin, padding;
  width: 0;
  height: 0;
  opacity: 0;
}
.dnb-form-status--is-animating {
  overflow: hidden;
  width: auto;
}
.dnb-form-status--disappear, .dnb-form-status--hidden {
  margin: 0 !important;
  padding: 0 !important;
}
.dnb-form-status__shell {
  display: flex;
  justify-content: flex-start;
  align-items: flex-start;
  min-width: inherit;
  border-radius: var(--form-status-radius);
}
.dnb-form-status__text {
  padding: 0.625rem 1rem;
  cursor: text;
  color: inherit;
  line-height: var(--line-height-small);
  font-size: var(--font-size-small);
  white-space: normal;
}
button .dnb-form-status__text {
  cursor: inherit;
}
.dnb-form-status__text .dnb-anchor {
  font-size: inherit;
}
.dnb-icon + .dnb-form-status__text {
  padding-left: 0.5rem;
}
.dnb-form-status__shell > .dnb-icon {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 0.3333333em 0.3333333em 0.3333333em 0.6666666em;
}
.dnb-form-status__size--large .dnb-form-status__text {
  padding-top: 1.125rem;
  padding-bottom: 1.125rem;
}
.dnb-form-status__size--large .dnb-form-status__shell > .dnb-icon {
  margin-top: 0.6666666em;
  margin-bottom: 0.6666666em;
}
.dnb-form-status--stretch {
  flex-grow: 1;
}
.dnb-form-status--stretch .dnb-form-status__shell {
  width: 100%;
}
.dnb-form-status--stretch .dnb-form-status__text {
  max-width: 47rem;
}
.dnb-form-status[hidden] {
  display: none;
}
.dnb-form-status--no-animation, html[data-visual-test] .dnb-form-status {
  transition-duration: 1ms !important;
}

/*
* Textarea component
*
*/
/*
 * Utilities
 */
:root {
  --textarea-padding-width: 0.5rem;
}

.dnb-textarea {
  display: inline-flex;
  font-size: var(--font-size-small);
  line-height: var(--line-height-basis);
}
.dnb-textarea__inner {
  display: inline-flex;
  flex-direction: column;
  margin: var(--textarea-padding-width);
}
.dnb-textarea__shell {
  display: inline-flex;
  position: relative;
  font-size: var(--font-size-basis);
}
.dnb-textarea__state {
  position: absolute;
  z-index: 1;
  top: -0.5rem;
  left: -0.5rem;
  bottom: -0.5rem;
  right: -0.5rem;
  background-color: var(--color-white);
  box-shadow: 0 0 0 0.0625rem var(--color-sea-green);
  border-radius: 0.25rem;
}
.dnb-textarea__row {
  display: inline-flex;
}
.dnb-textarea__suffix {
  padding-left: 1.5rem;
}
.dnb-textarea__textarea {
  position: relative;
  z-index: 2;
  cursor: auto;
  border: none;
  margin: 0;
  padding: 0 var(--textarea-padding-width);
  outline: none;
  font-size: var(--font-size-basis);
  line-height: var(--line-height-basis);
}
.dnb-textarea__placeholder {
  pointer-events: none;
  position: absolute;
  z-index: 3;
  top: 0;
  left: 0;
  overflow: hidden;
  width: 100%;
  height: 100%;
  cursor: auto;
  border: none;
  margin: 0;
  padding: 0 var(--textarea-padding-width);
  outline: none;
  font-size: var(--font-size-basis);
  line-height: var(--line-height-basis);
}
.dnb-textarea__textarea, .dnb-textarea__placeholder {
  text-align: left;
}
.dnb-textarea__align--right .dnb-textarea__textarea, .dnb-textarea__align--right .dnb-textarea__placeholder {
  text-align: right;
}
.dnb-textarea__autoresize .dnb-textarea__textarea {
  resize: none;
}
.dnb-textarea--has-content .dnb-textarea__textarea ~ .dnb-textarea__placeholder, .dnb-textarea--focus .dnb-textarea__textarea:not([disabled]):not([readonly]) ~ .dnb-textarea__placeholder {
  opacity: 0.5;
}
.dnb-textarea__inner > .dnb-form-status {
  order: 2;
  margin: 1rem 0 0 calc(0px - var(--textarea-padding-width));
}
.dnb-textarea:not(.dnb-textarea--vertical) .dnb-form-label {
  margin-top: 0.5rem;
}
.dnb-textarea--vertical {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}
.dnb-textarea:not(.dnb-textarea--vertical)[class*=__status] {
  align-items: flex-start;
}
.dnb-textarea:not(.dnb-textarea--vertical)[class*=__status] > .dnb-form-label {
  margin-top: 0.25rem;
}
@media screen and (max-width: 40em) {
  .dnb-textarea {
    flex-wrap: wrap;
  }
  .dnb-textarea > .dnb-form-label {
    margin-bottom: 0.5rem;
    margin-top: 0.5rem;
  }
}
.dnb-textarea--stretch {
  display: flex;
  flex-grow: 1;
}
.dnb-textarea--stretch .dnb-textarea__inner {
  flex-grow: 1;
}
.dnb-textarea--stretch .dnb-textarea__shell, .dnb-textarea--stretch .dnb-textarea__inner, .dnb-textarea--stretch .dnb-textarea__textarea {
  width: 100%;
}
.dnb-textarea--stretch .dnb-form-label + .dnb-textarea__inner {
  width: auto;
}
.dnb-textarea--vertical.dnb-textarea--stretch .dnb-textarea__inner {
  width: 100%;
}
.dnb-form-row--horizontal .dnb-textarea--stretch {
  width: 100%;
}
@media screen and (max-width: 40em) {
  .dnb-responsive-component .dnb-textarea {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    margin-bottom: 0.5rem;
  }
  .dnb-responsive-component .dnb-textarea > .dnb-form-label {
    margin-bottom: 0.5rem;
  }
}
html[data-visual-test] .dnb-textarea__textarea {
  caret-color: var(--color-white);
}
.dnb-textarea.dnb-skeleton .dnb-textarea__inner {
  margin: 0;
  border-radius: 0.25rem;
}
.dnb-textarea.dnb-skeleton .dnb-textarea__inner .dnb-textarea__shell {
  margin: var(--textarea-padding-width);
}
.dnb-textarea.dnb-skeleton .dnb-textarea__state, .dnb-textarea.dnb-skeleton .dnb-textarea__textarea {
  visibility: hidden;
}"
`;
